src/Data/Aeson/Types/Internal.hs

```diff
@@ -306,6 +308,15 @@ instance Monad.Monad Parser where
     {-# INLINE fail #-}
 #endif
 
+-- |
+--
+-- @since 2.1.0.0
+instance MonadFix Parser where
+    mfix f = let p = f (lower [] p) in p
```

Comment:
The fixed point should not be a function and the `path` should be taken from the `Parser` context.

```
-- I haven't typechecked this
mfix f = Parser $ \path kf ks -> let x = runParser (f (fromISuccess x)) path IError ISuccess in
  case x of
    IError p e -> kf p e
    ISuccess y -> ks y

fromISuccess :: IResult a -> a
```

Reply:
But what's the difference. Can we write a test which differentiates this and the current impl?

Reply:
`take 3 <$> mfix (\x -> pure (trace "x" (1 : x)))` should trace only once.

Reply:
Both implementations trace only once, AFAICT. (By trying in ghci).

Reply:
Anyway, yours has `path` thing, so it's better. Let's go with it.

Reply:
Oops it's a bit more subtle than I thought, but here's a working litmus test:

```
>>> parse (\_ -> take 3 <$> mfix (\x -> return () >>= \_ -> pure (trace "THIS SHOULD PRINT ONLY ONCE" (() : x)))) ()
Success THIS SHOULD PRINT ONLY ONCE
[()THIS SHOULD PRINT ONLY ONCE
,(),()]
```